code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/AEPNotificationUtil.kt

```diff
@@ -0,0 +1,150 @@
+/*
+  Copyright 2024 Adobe. All rights reserved.
+  This file is licensed to you under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile.services.ui.notification
+
+import android.app.Activity
+import android.content.BroadcastReceiver
+import android.content.Intent
+import androidx.core.app.NotificationCompat
+import com.adobe.marketing.mobile.services.Log
+import com.adobe.marketing.mobile.services.ServiceProvider
+import com.adobe.marketing.mobile.services.ui.notification.builders.AutoCarouselNotificationBuilder
+import com.adobe.marketing.mobile.services.ui.notification.builders.BasicNotificationBuilder
+import com.adobe.marketing.mobile.services.ui.notification.builders.LegacyNotificationBuilder
+import com.adobe.marketing.mobile.services.ui.notification.builders.ManualCarouselNotificationBuilder
+import com.adobe.marketing.mobile.services.ui.notification.models.AEPPushTemplate
+import com.adobe.marketing.mobile.services.ui.notification.models.AutoCarouselPushTemplate
+import com.adobe.marketing.mobile.services.ui.notification.models.BasicPushTemplate
+import com.adobe.marketing.mobile.services.ui.notification.models.CarouselPushTemplate
+import com.adobe.marketing.mobile.services.ui.notification.models.ManualCarouselPushTemplate
+
+/**
+ * Public facing object to construct a [NotificationCompat.Builder] object for the specified [PushTemplateType].
+ * The [constructNotificationBuilder] methods will build the appropriate notification based on the provided
+ * [AEPPushTemplate] or [Intent].
+ */
+object AEPNotificationUtil {
+    private const val SELF_TAG = "AepNotificationUtil"
+
+    @Throws(NotificationConstructionFailedException::class)
+    @JvmStatic
+    fun constructNotificationBuilder(
+        messageData: Map<String, String>,
+        trackerActivityClass: Class<out Activity>?,
+        broadcastReceiverClass: Class<out BroadcastReceiver>?
+    ): NotificationCompat.Builder {
+        val context = ServiceProvider.getInstance().appContextService.applicationContext
+            ?: throw NotificationConstructionFailedException("Application context is null, cannot build a notification.")
+        val pushTemplateType =
+            PushTemplateType.fromString(messageData[PushTemplateConstants.PushPayloadKeys.TEMPLATE_TYPE])
+
+        when (pushTemplateType) {
+            PushTemplateType.BASIC -> {
+                val basicPushTemplate = BasicPushTemplate(messageData as MutableMap<String, String>)
```

Comment:
This is not a safe cast. Why do the templates need a MutableMap? Would it be acceptable to use Map instead?

Reply:
Doesn't need to be MutableMap, its only read from. Updating it use Map